### DISCUSSION TITLE
DPCTGAN

### DISCUSSION THREAD

**Discussion by gpadres** [opendp/opendp#1241] [Differential Privacy (Theory/Practice)] [answered]
Hi is there documentation or further literature on DPCTGAN?
I found the following paper https://ml-research.github.io/papers/fang2022dpctgan.pdf but their implemented code seems very different from OpenDP implementation.
Thanks in advance

**Comment by Shoeboxam** [comment]
Tagging @joshua-oss!

**Comment by joshua-oss** [comment, marked as answer]
Their implementation appears to be quite similar to ours: We just took the CTGAN paper and implemented with DPSGD using Opacus.

I've been thinking of deprecating ours, because utility suffered greatly after we fixed some privacy bugs. It looks like their implementation may have the two privacy bugs that accounted for most of our utility. When sampling the conditional vector, CTGAN operates by oversampling from rows that have categories that are more rare (with log sampling). This is the key piece that prevents mode collapse. 

We couldn't find a way to make that private, so we eliminated that code and sample only uniformly. Their implementation defaults to log frequency, but they have an option that supports this same frequency-based sampling we reverted to, though they use exact counts instead of noisy counts. In any case this turns into uniform random sampling and eliminates much of the utility.

They also preprocess with BayesianGMM, which is not differentially private. This is borrowed from the original CTGAN paper and is what we did at first. Removing that had a negative impact on utility.

https://github.com/opendp/smartnoise-sdk/issues/484
